field.go

```diff
@@ -56,11 +69,31 @@ func (f *field) SourceCodeInfo() SourceCodeInfo               { return f.info }
 func (f *field) Descriptor() *descriptor.FieldDescriptorProto { return f.desc }
 func (f *field) Message() Message                             { return f.msg }
 func (f *field) InOneOf() bool                                { return f.oneof != nil }
+func (f *field) InRealOneOf() bool                            { return f.InOneOf() && !f.desc.GetProto3Optional() }
 func (f *field) OneOf() OneOf                                 { return f.oneof }
 func (f *field) Type() FieldType                              { return f.typ }
 func (f *field) setMessage(m Message)                         { f.msg = m }
 func (f *field) setOneOf(o OneOf)                             { f.oneof = o }
 
+func (f *field) HasPresence() bool {
+	if f.Type().IsRepeated() || f.Type().IsMap() {
+		return false
+	}
+
+	if f.Syntax() == Proto3 && !f.Optional() && !(f.Type().IsEmbed() || f.InRealOneOf()) {
+		return false
+	}
+
+	return true
+}
+
+func (f *field) Optional() bool {
+	if f.Syntax() == Proto3 {
```

Comment:
Both `proto2` & `proto3` support explicit presence.

I can see why you would say that, but since presence here is a property of a field and not any particular syntax, that function wouldn't be particularly useful or accurate imo.

Reply:
Agree, thanks.